### PR TITLE
[CMake] Handle clang in MSVC mode in GetHostTriple

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
When configuring CMake with Clang in MSVC mode, Clang can either be invoked with the MSVC style driver (clang-cl) or the GNU style driver (clang). When using the MSVC style driver, CMake sets the CMake variable MSVC (which indicates the kind of command line interface), but when using the GNU style driver, this variable isn't set, while Clang still operates in MSVC mode.

Even though CMake doesn't set the variable MSVC, it still does set CMAKE_C_COMPILER_ARCHITECTURE_ID, which it does set for MSVC and Clang in MSVC mode, but not for Clang in MinGW mode.

For this configuration, use the MSVC style, CMAKE_C_COMPILER_ARCHITECTURE_ID based GetHostTriple implementation.

### CODE REVIEW

File: llvm/cmake/modules/GetHostTriple.cmake [hunk 1]
```diff
@@ -42,7 +42,21 @@ function( get_host_triple var )
...
+      elseif( CMAKE_C_COMPILER_ARCHITECTURE_ID MATCHES "ARM.*" )
+        set( value "armv7-pc-windows-msvc" )
+      elseif( CMAKE_C_COMPILER_ARCHITECTURE_ID STREQUAL "x64" )
+        set( value "x86_64-pc-windows-msvc" )
+      elseif( CMAKE_C_COMPILER_ARCHITECTURE_ID STREQUAL "X86" )
+        set( value "i686-pc-windows-msvc" )
+      elseif( CMAKE_SIZEOF_VOID_P EQUAL 8 )
+        set( value "x86_64-pc-windows-msvc" )
+      else()
+        set( value "i686-pc-windows-msvc" )
```
Comment (petrhosek) [inline llvm/cmake/modules/GetHostTriple.cmake:57]: We usually don't use spaces around parens.
```suggestion
      if(CMAKE_C_COMPILER_ARCHITECTURE_ID MATCHES "ARM64.*")
        set(value "aarch64-pc-windows-msvc")
      elseif(CMAKE_C_COMPILER_ARCHITECTURE_ID MATCHES "ARM.*")
        set(value "armv7-pc-windows-msvc")
      elseif(CMAKE_C_COMPILER_ARCHITECTURE_ID STREQUAL "x64")
        set(value "x86_64-pc-windows-msvc")
      elseif(CMAKE_C_COMPILER_ARCHITECTURE_ID STREQUAL "X86")
        set(value "i686-pc-windows-msvc")
      elseif(CMAKE_SIZEOF_VOID_P EQUAL 8)
        set(value "x86_64-pc-windows-msvc")
      else()
        set(value "i686-pc-windows-msvc")
```

Comment (petrhosek) [inline llvm/cmake/modules/GetHostTriple.cmake:57]: Should we assume x86/x64 when `CMAKE_C_COMPILER_ARCHITECTURE_ID` isn't `x86` or `x64`? Can we report an error instead?

File: llvm/cmake/modules/GetHostTriple.cmake [hunk 2]
```diff
@@ -42,7 +42,21 @@ function( get_host_triple var )
       set( value "powerpc-ibm-aix" )
     endif()
   else( MSVC )
-    if(CMAKE_HOST_SYSTEM_NAME STREQUAL Windows AND NOT MSYS)
+    if(CMAKE_HOST_SYSTEM_NAME STREQUAL Windows AND NOT MSYS AND CMAKE_C_COMPILER_ID MATCHES "Clang" )
```
Comment (mstorsjo) [inline llvm/cmake/modules/GetHostTriple.cmake:45]: Why is this needed here? When operating in MSVC mode, CMake usually sets the `MSVC` CMake variable, so this should hit the existing codepath at lines 5-18 above? And this is a bit weird, because CMake usually doesn't set `COMPILER_ARCHITECTURE_ID` for Clang (or GCC), but only for MSVC. I guess CMake might set `COMPILER_ARCHITECTURE_ID` for Clang-cl when operating in MSVC mode, but if we're invoking it as `clang`, not `clang-cl` so that the `MSVC` variable isn't set, it feels odd that CMake still would be setting `COMPILER_ARCHITECTURE_ID`.

Also I don't quite see why this needs to be as a separate step within the `else()` clause here, why can't this be a toplevel case like the other ones?

The condition, `CMAKE_HOST_SYSTEM_NAME STREQUAL Windows AND NOT MSYS` also feels quite contrieved for this case. I guess what you want to detect is "clang, targeting windows, probably in msvc mode (as we're past a  check for `MINGW` further up)", and we generally don't want to check for the host unless we need to.

The existing check for `CMAKE_HOST_SYSTEM_NAME STREQUAL Windows AND NOT MSYS` is to pick out configurations where we _can't_  run the generic detection shell script below.


Can't we just extend the existing MSVC case at the top, e.g. something like `if (MSVC OR (CMAKE_C_COMPILER_ID MATCHES "Clang" AND WIN32 AND CMAKE_C_COMPILER_ARCHITECTURE_ID)` or something along those lines, to specifically match the case we're looking for here?

**Review by mstorsjo — COMMENTED** [review]
The PR subject/description reads a bit weird, can you rephrase it?

I guess "build clang with..." means "make sure that building clang works", but it feels a bit terse and off the point here, I'd rather focus on e.g. "fix GetHostTriple for clang in MSVC mode, with the GCC like command line driver mode" or something like that?

(Also, the second sentence in the PR description feels a bit off, that could also use a couple more words to be more readable.)

File: llvm/cmake/modules/GetHostTriple.cmake [hunk 3]
```diff
@@ -2,7 +2,7 @@
 # Invokes config.guess
 
 function( get_host_triple var )
-  if( MSVC )
+  if(MSVC OR (CMAKE_HOST_SYSTEM_NAME STREQUAL Windows AND CMAKE_C_COMPILER_ID MATCHES "Clang"))
```
Comment (mstorsjo) [inline llvm/cmake/modules/GetHostTriple.cmake:5]: This isn't about `CMAKE_HOST_SYSTEM_NAME`, it's about the target, so we should check `CMAKE_SYSTEM_NAME` instead - to make it work as expected and intended for cross compilation.

The existing check below which checks `CMAKE_HOST_SYSTEM_NAME` is for determining whether it will be possible to execute a shell script, where we do need to check the host environment.

**Comment by mstorsjo** [comment]
The commit message still reads very very weirdly to me - it talks a lot about goals and other stuff, instead of focusing on the actual CMake configuration. (All the other aspects on what tools are available and which sets of tests pass in this configuration, are entirely unrelated to this change.)

May I suggest this instead, which focues on the thing at hand?

---

Subject: [CMake] Handle clang in MSVC mode in GetHostTriple

When configuring CMake with Clang in MSVC mode, Clang can either be invoked with the MSVC style driver (clang-cl) or the GNU style driver (clang). When using the MSVC style driver, CMake sets the CMake variable `MSVC` (which indicates the kind of command line interface), but when using the GNU style driver, this variable isn't set, while Clang still operates in MSVC mode.

Even though CMake doesn't set the variable `MSVC`, it still does set `CMAKE_C_COMPILER_ARCHITECTURE_ID`, which it does set for MSVC and Clang in MSVC mode, but not for Clang in MinGW mode.

For this configuration, use the MSVC style, `CMAKE_C_COMPILER_ARCHITECTURE_ID` based GetHostTriple implementation.

---

Doesn't that capture the thing that this patch actually deals with? If I read the code correctly, `GetHostTriple` would fail in the same way also on x86, if CMake is invoked with Clang in MSVC mode, with the GNU style driver?

File: llvm/cmake/modules/GetHostTriple.cmake [hunk 4]
```diff
@@ -2,7 +2,7 @@
 # Invokes config.guess
 
 function( get_host_triple var )
-  if( MSVC )
+  if( MSVC OR (CMAKE_SYSTEM_NAME STREQUAL "Windows" AND CMAKE_C_COMPILER_ID MATCHES "Clang" AND NOT MINGW AND NOT MSYS))
```
Comment (omjavaid) [inline llvm/cmake/modules/GetHostTriple.cmake:5]: IMO, this if condition should be organized something like following:
```cmake
if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
    if(CMAKE_C_COMPILER_ID MATCHES "Clang")
        if(MSVC)
            message(STATUS "Clang is operating in MSVC mode (MSVC style driver: clang-cl).")
        elseif(CMAKE_C_COMPILER_ARCHITECTURE_ID)
            message(STATUS "Clang is operating in MSVC mode (GNU style driver).")
        else()
            message(STATUS "Clang MinGW mode")
        endif()
    elseif(MSVC)
        message(STATUS "Compiler is MSVC (cl compiler).")
endif()

Comment (mstorsjo) [reply llvm/cmake/modules/GetHostTriple.cmake:5]: Yeah, something like that would certainly be cleaner - but that would probably also be a much larger refactoring of the cmake code here.

Anyway, I'll accept the change here as I think this version should function correctly, but I don't mind such a restructuring either (maybe ideally as a separate step?).

**Review by mstorsjo — APPROVED** [review]
LGTM, I think this version works as wanted.

File: llvm/cmake/modules/GetHostTriple.cmake (hunk 4 above)
Comment (DanielKristofKiss) [reply llvm/cmake/modules/GetHostTriple.cmake:5]: I'll take care of that later, just first want to be sure all changes there to run all test, build everything fine. Not sure how much other changes needed... looking at it.

**Comment by mstorsjo** [comment]
Btw, there's another way of detecting Clang in MSVC mode, while using the GNU command line. Currently we do this as `(WIN32 AND NOT MINGW)` or similar, but there's also `CMAKE_CXX_SIMULATE_ID STREQUAL "MSVC"` which should trigger for both `clang` (in MSVC mode) and `clang-cl`, but not `clang` in mingw mode. See https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_SIMULATE_ID.html for docs, and https://discourse.cmake.org/t/clang-on-windows-detected-with-msvc-interface/3470/3 for a concrete example.

**Comment by DanielKristofKiss** [comment]
@mstorsjo  addressed in #122914
